rinja_parser/src/lib.rs

```diff
@@ -363,25 +363,78 @@ fn ws<'a, O>(
 /// Skips input until `end` was found, but does not consume it.
 /// Returns tuple that would be returned when parsing `end`.
 fn skip_till<'a, O>(
+    candidate_finder: impl IntoSkipTillCandidateFinder,
     end: impl FnMut(&'a str) -> ParseResult<'a, O>,
 ) -> impl FnMut(&'a str) -> ParseResult<'a, (&'a str, O)> {
-    enum Next<O> {
-        IsEnd(O),
-        NotEnd,
-    }
-    let mut next = alt((map(end, Next::IsEnd), map(anychar, |_| Next::NotEnd)));
+    let candidate_finder = candidate_finder.into_candidate_finder();
+    let mut next = alt((map(end, Some), map(anychar, |_| None)));
     move |start: &'a str| {
         let mut i = start;
         loop {
-            let (j, is_end) = next(i)?;
-            match is_end {
-                Next::IsEnd(lookahead) => return Ok((i, (j, lookahead))),
-                Next::NotEnd => i = j,
-            }
+            i = match candidate_finder.next(i) {
+                Some(j) => j,
+                None => return Err(nom::Err::Error(ErrorContext::new("`end` not found`", i))),
+            };
+            i = match next(i)? {
+                (j, Some(lookahead)) => return Ok((i, (j, lookahead))),
+                (j, None) => j,
+            };
         }
     }
 }
 
+trait IntoSkipTillCandidateFinder: Copy {
+    type CandidateFinder: CandidateFinder;
+
+    fn into_candidate_finder(self) -> Self::CandidateFinder;
+}
+
+trait CandidateFinder: Copy {
+    fn next(self, haystack: &str) -> Option<&str>;
+}
+
+/// Implement `CandidateFinder` for different count of needles.
+macro_rules! impl_candidate_finder {
+    ($(($Ty:ident $count:literal $($field:ident)+))*) => { $(
+        // Hiding implementation details for the `CandidateFinder`s, so we can be sure that we
+        // operate on valid `&str`s / initial UTF-8 bytes.
+        const _: () = {
+            use memchr::arch::all::memchr::$Ty;
+
+            #[derive(Clone, Copy)]
+            struct CandidateFinderN($Ty);
+
+            impl IntoSkipTillCandidateFinder for [&str; $count] {
+                type CandidateFinder = CandidateFinderN;
+
+                #[inline]
+                fn into_candidate_finder(self) -> CandidateFinderN {
+                    let [$($field),+] = self;
+                    CandidateFinderN($Ty::new($($field.as_bytes()[0]),+))
+                }
+            }
+
+            impl CandidateFinder for CandidateFinderN {
+                #[inline]
+                fn next(self, haystack: &str) -> Option<&str> {
+                    let haystack = haystack.as_bytes();
+                    let i = self.0.find(haystack)?;
+                    // SAFETY: We know that `memchr` returns a valid position:
+                    // The needles can only be ASCII, or the start of a UTF-8 sequence, so
+                    // only an ASCII character or the start of a UTF-8 sequence can be follow.
```

Comment:
"can be follow"?

Reply:
"Can be" that I should read my comments again after writing them. :)

Reply:
That's what reviews are meant for. :D